sematic/log_reader.py

```diff
@@ -340,16 +341,33 @@ def load_log_lines(
             reverse=reverse,
             **cursor_kwargs,  # type: ignore
         )
+    else:
+        log_line_result = _load_non_inline_logs(
+            run_id=run_id,
+            still_running=still_running,
+            max_lines=max_lines,
+            filter_strings=filter_strings,
+            default_log_info_message=default_log_info_message,
+            reverse=reverse,
+            **cursor_kwargs,  # type: ignore
+        )
 
-    return _load_non_inline_logs(
-        run_id=run_id,
-        still_running=still_running,
-        max_lines=max_lines,
-        filter_strings=filter_strings,
-        default_log_info_message=default_log_info_message,
-        reverse=reverse,
-        **cursor_kwargs,  # type: ignore
-    )
+    # when no cursor is set, it means that we are doing the initial pull
+    # and we can determine if we can continue in either direction
+    # this overrides the value from the `log_line_result`.
+    if reverse_cursor_token is None and forward_cursor_token is None:
+        if reverse is True:
+            # can_continue_forward will be False
+            # if the run has reached to a terminal state
+            log_line_result = replace(
+                log_line_result,
+                can_continue_forward=run.future_state
+                not in FutureState.terminal_state_strings(),
+            )
+        else:
+            log_line_result = replace(log_line_result, can_continue_backward=False)
```

Comment:
This logic should *probably* be pushed all the way down to `get_log_lines_from_line_stream`, but I won't block on it.

Reply:
Yes, probably. I evaluated the two, the current way might involve less code complexity. `get_log_lines_from_line_stream` need to deal with getting the run's future_state (either passed in or get_run() again). At the end of the day, the moment this logic is triggered will be the same. So when balancing the semantics and complexity, I chose the lesser of the two evils. If you feel this significantly distorts the code flow, I can look into pushing it to `get_log_lines_from_line_stream`.